ksqldb-functional-tests/src/test/resources/query-validation-tests/fk-join.json

```diff
@@ -268,6 +268,417 @@
+        {"topic": "right_topic", "key": 0, "value": {"F1": "a", "F2": 10}, "timestamp": 15000},
+        {"topic": "left_topic", "key": {"L_ID": 10, "L_ID_2": 0}, "value": {"NAME": "bar", "VALUE": 0}, "timestamp": 16000},
+        {"topic": "right_topic", "key": 0, "value": null, "timestamp": 17000},
+        {"topic": "left_topic", "key": {"L_ID": 1, "L_ID_2": 0}, "value": null, "timestamp": 18000},
+        {"topic": "left_topic", "key": {"L_ID": 0, "L_ID_2": 100}, "value": null, "timestamp": 19000}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "key": {"L_ID": 1, "L_ID_2": 0}, "value": {"R_ID": null, "NAME": "zero", "F1": null}, "timestamp": 0},
+        {"topic": "OUTPUT", "key": {"L_ID": 1, "L_ID_2": 0}, "value": {"R_ID": 0, "NAME": "zero", "F1": "blah"}, "timestamp": 11000},
+        {"topic": "OUTPUT", "key": {"L_ID": 0, "L_ID_2": 100}, "value": {"R_ID": null, "NAME": "foo", "F1": null}, "timestamp": 13000},
+        {"topic": "OUTPUT", "key": {"L_ID": 10, "L_ID_2": 0}, "value": {"R_ID": 0, "NAME": "bar", "F1": "a"}, "timestamp": 16000},
+        {"topic": "OUTPUT", "key": {"L_ID": 1, "L_ID_2": 0}, "value": null, "timestamp": 18000},
+        {"topic": "OUTPUT", "key": {"L_ID": 0, "L_ID_2": 100}, "value": null, "timestamp": 19000}
+      ],
+      "post": {
+        "sources": [
+          {"name": "OUTPUT", "type": "table", "schema": "L_ID BIGINT KEY, L_ID_2 BIGINT KEY, `R_ID` STRING, `NAME` STRING, `F1` STRING"}
+        ]
+      }
+    },
+    {
+      "name": "fails on partial left source key",
+      "properties": {
+        "ksql.joins.foreign.key.enable": true
+      },
+      "statements": [
+        "CREATE TABLE left_table (L_ID BIGINT PRIMARY KEY, L_ID_2 BIGINT PRIMARY KEY, NAME varchar, VALUE bigint) WITH (kafka_topic='left_topic', format='JSON');",
+        "CREATE TABLE right_table (R_ID BIGINT PRIMARY KEY, F1 varchar, F2 bigint) WITH (kafka_topic='right_topic', format='JSON');",
+        "CREATE TABLE OUTPUT as SELECT l_id, r_id, name, f1 FROM left_table JOIN right_table ON value = r_id;"
+      ],
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlStatementException",
+        "message": "TODO: some sensible error message here"
+      }
+    },
+    {
+      "name": "fails on type mismatch",
+      "properties": {
+        "ksql.joins.foreign.key.enable": true
+      },
+      "statements": [
+        "CREATE TABLE left_table (L_ID BIGINT PRIMARY KEY, NAME varchar, VALUE bigint) WITH (kafka_topic='left_topic', value_format='JSON');",
+        "CREATE TABLE right_table (R_ID BIGINT PRIMARY KEY, F1 varchar, F2 bigint) WITH (kafka_topic='right_topic', value_format='JSON');",
+        "CREATE TABLE OUTPUT as SELECT l_id, r_id, name, f1 FROM left_table JOIN right_table ON name = r_id;"
+      ],
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlStatementException",
+        "message": "TODO: some sensible error message here"
```

Comment:
This error message will be added in https://github.com/confluentinc/ksql/pull/7517 (or possible in https://github.com/confluentinc/ksql/pull/7522, based on the outcome of the discussion in https://github.com/confluentinc/ksql/pull/7517#discussion_r631984712).